### DISCUSSION TITLE
two realsense D435 cameras on a thinkpad laptop with 3080 GPU for object detection using YoloV10 (racing for GPU and getting stuck after first detection)

### DISCUSSION THREAD

**Discussion by monajalal** [ultralytics/ultralytics#16018] [General]
I have no problem with having one single camera doing object detection using yolov10 (Python).

I also have no problem with getting live RGB feed from both cameras.

However, when I connect both cameras and do object detection using Ultralytics yolov10, it crashes after detecting an object.

Please note, I am showing results in two cv2 named windows, RealSense1 and RealSense2.

I understand there is an apparent race for GPU but how can I resolve this for when I have two live RGB feeds from two different Intel RealSense D435 cameras?

I did see this https://github.com/ultralytics/ultralytics/issues/5810#issuecomment-2286561321 by @glenn-jocher  but cannot figure how to make it work for video stream (vs an offline video).

Do you have any code sample or pseudo code for this?

I have ubuntu 22.04 on my machine.

Currently, each RGB camera runs at 30 FPS.

**Comment by glenn-jocher** [comment]
@monajalal it sounds like a resource contention issue with the GPU. Try processing each camera feed in separate threads or processes to manage GPU usage more effectively. You can use Python's `threading` or `multiprocessing` modules to achieve this. Ensure you're using the latest version of the Ultralytics package, as updates may resolve performance issues.

**Comment by monajalal** [comment]
@glenn-jocher 

I have this code, and after I display the results of camera 1 on the first display, and it does the prediction, it crashes (gets stuck). Could you please help me with fixing this?

I have two Intel RealSense D435 camera, two different displays (besides the laptop display), and one 3080 GPU inside the laptop.

```

pred_queue1 = queue.Queue()
pred_queue2 = queue.Queue()

frame_queue = queue.Queue(maxsize=2)
lock = threading.Lock()

def prediction_worker1(model, frame):
    result_img = predict_and_detect(model, frame, [0, 7], 0.5)
    pred_queue1.put(result_img)
    
    
def prediction_worker2(model, frame):
    result_img = predict_and_detect(model, frame, [0, 7], 0.5)
    pred_queue2.put(result_img)
    
    
    
def predict(chosen_model, img, classes=[], conf=0.5):
    if classes:
        results = chosen_model.predict(img, classes=classes, conf=conf)
    else:
        results = chosen_model.predict(img, conf=conf)

    return results

def predict_and_detect(chosen_model, img, classes=[], conf=0.5, rectangle_thickness=3, text_thickness=3):
    results = predict(chosen_model, img, classes, conf=conf)
    for result in results:
        for box in result.boxes:
            cv2.rectangle(img, (int(box.xyxy[0][0]), int(box.xyxy[0][1])),
                          (int(box.xyxy[0][2]), int(box.xyxy[0][3])), (49, 49, 255), rectangle_thickness)
            cv2.putText(img, f"{result.names[int(box.cls[0])]}",
                        (int(box.xyxy[0][0]), int(box.xyxy[0][1]) - 10),
                        cv2.FONT_HERSHEY_SIMPLEX, 1, (49, 49, 255), text_thickness)
    return img, results




model = YOLO("yolov10x.pt")

def sync_capture(frame_queue, lock, camera_SN):
    pipeline = rs.pipeline()
    config = rs.config()
    config.enable_stream(rs.stream.color, 1920, 1080, rs.format.bgr8, 30)
    config.enable_device(f"{camera_SN}")
    pipeline.start(config)
    while True:
        frames = pipeline.wait_for_frames()
        color_frame = frames.get_color_frame()
        if not color_frame:
            continue
        
        color_image = np.asanyarray(color_frame.get_data())
        
        with lock:
            frame_queue.put((camera_SN, color_image))
    

def process_frames(model, frame_queue, lock):
    while True:
        with lock:
            if not frame_queue.empty():
                camera_SN, frame = frame_queue.get()
        
                if camera_SN == SN1:
                    cv2.namedWindow("RealSense1", cv2.WINDOW_NORMAL)
                    cv2.moveWindow("RealSense1", primary_display_x-1, primary_display_y-1) 
           
                    #result_img, _ = predict_and_detect(model, frame, classes=[0, 7], conf=0.5) #doesn't work
                    prediction_thread1 = threading.Thread(target=prediction_worker1, args=(model, frame))
                    prediction_thread1.start()
                    prediction_thread1.join()
                    try:
                        result_img = pred_queue1.get(timeout=1)[0]
                        print('type result_img 1: ', type(result_img))
                        print('result_img 1: ', result_img)
                        cv2.imshow('RealSense1', result_img)
                    except pred_queue1.empty():
                        print("prediction queue is empty")
                    
                elif camera_SN == SN2:
                    cv2.namedWindow("RealSense2", cv2.WINDOW_NORMAL) 
                    cv2.moveWindow("RealSense2", second_display_x-1, second_display_y-1) 
                  
                    # result_img, _ = predict_and_detect(model, frame, classes=[0, 7], conf=0.5) # doesn't work
                    prediction_thread2 = threading.Thread(target=prediction_worker1, args=(model, frame))
                    prediction_thread2.start()
                    prediction_thread2.join()
                    try:
                        result_img = pred_queue2.get(timeout=1)[0]
                        print('type result_img 2: ', type(result_img))
                        print('result_img 2: ', result_img)
                        cv2.imshow('RealSense2', result_img)
                    except pred_queue2.empty():
                        print("pred queue 2 empty")
                    
            
        if cv2.waitKey(1) == ord('q'):
            break
        torch.cuda.empty_cache()
        
        
if __name__ == '__main__':

    camera_thread1 = threading.Thread(target=sync_capture, args=(frame_queue, lock, SN1))
    camera_thread2 = threading.Thread(target=sync_capture, args=(frame_queue, lock, SN2))
    process_thread = threading.Thread(target=process_frames, args=(model, frame_queue, lock))
    
    camera_thread1.start()
    camera_thread2.start()
    process_thread.start()
    
    camera_thread1.join()
    camera_thread2.join()
    process_thread.join()

    cv2.destroyAllWindows()

```

**Reply by glenn-jocher** [reply]
@monajalal it looks like the issue might be related to threading and GPU resource contention. Here are a few suggestions to help resolve the problem:

1. **Avoid Blocking Threads**: Instead of using `join()` immediately after starting threads, consider using a queue to manage frame processing asynchronously. This will prevent blocking and allow both cameras to process frames concurrently.

2. **Separate GPU Contexts**: Ensure that each thread has its own GPU context. You can achieve this by setting the CUDA device within each thread using `torch.cuda.set_device()`.

3. **Optimize Queue Usage**: Use a single queue for frames and results, and ensure that the queue operations are thread-safe to avoid race conditions.

4. **Check for Empty Queues**: Use `queue.Empty` exception handling correctly to manage empty queues.

5. **Update to Latest Version**: Ensure you are using the latest version of the Ultralytics package to benefit from any recent bug fixes or improvements.

If the issue persists, consider simplifying the code to isolate the problem, and test with a single camera setup first. For more detailed guidance, you might want to explore the Ultralytics documentation or community forums.

**Comment by monajalal** [comment]
@glenn-jocher
Thanks a lot for your response.

I am a bit confused how
Separate GPU Contexts: Ensure that each thread has its own GPU context. You can achieve this by setting the CUDA device within each thread using torch.cuda.set_device().

could help me if **I only have one GPU**. Could you please elaborate further?

**Reply by glenn-jocher** [reply]
@monajalal to handle multiple camera feeds with a single GPU, ensure each thread initializes its own YOLO model instance. This isolates GPU contexts and prevents conflicts. For more details, check our thread-safe inference guide: https://docs.ultralytics.com/guides/yolo-thread-safe-inference/.
